src/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs

```diff
@@ -73,39 +88,54 @@ public void Reset()
             }
         }
 
-        public bool IsCompleted => ReferenceEquals(_state, _awaitableIsCompleted);
-        internal bool HasContinuation => !ReferenceEquals(_state, _awaitableIsNotCompleted);
+        public bool IsCompleted => ReferenceEquals(_completion, _awaitableIsCompleted);
+        internal bool HasContinuation => !ReferenceEquals(_completion, _awaitableIsNotCompleted);
 
-        public Action OnCompleted(Action continuation, out bool doubleCompletion)
+        public void OnCompleted(Action<object> continuation, object state, ValueTaskSourceOnCompletedFlags flags, out Action<object> completion, out object completionState, out bool doubleCompletion)
         {
+            completionState = null;
+            completion = null;
+
             doubleCompletion = false;
-            Action awaitableState = _state;
+            Action<object> awaitableState = _completion;
             if (ReferenceEquals(awaitableState, _awaitableIsNotCompleted))
             {
-                _state = continuation;
+                _completion = continuation;
+                _completionState = state;
+
+                if ((flags & ValueTaskSourceOnCompletedFlags.UseSchedulingContext) != 0)
+                {
+                    // Set the scheduler to the current synchronization context if there is one
+                    // otherwise we delegate to what the pipe was configured with.
+                    // REVIEW: Should the sync context override the current scheduler if it was explicitly specifed
+                    // in the pipe options?
+                    if (SynchronizationContext.Current != null)
+                    {
+                        _scheduler = PipeScheduler.SynchronizationContext;
```

Comment:
I don't think this is going to do the right thing.  It looks like the SynchronizationContextPipeScheduler is looking at Current at the time it does the scheduling, rather than here.  But the whole point of capturing the context is to grab it at the point of the await so that it can be used to return to the right context when the operation completes asynchronously somewhere that isn't already on original context.  By having the scheduling do SynchronizationContext.Current.Post at the time of operation completion, it's not going to be going to the right place.

Reply:
Yes you’re right, I was trying to avoid the allocation here but I’ll have to allocate. Maybe I’ll have to allocate or just store the sync context itself and return that instead of a pipe scheduler.